File: binding/ios/Cheetah.swift
//
//  Copyright 2022-2026 Picovoice Inc.
//  You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
//  file accompanying this source.
//  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
//  specific language governing permissions and limitations under the License.
//

import Foundation

import PvCheetah
import Yams

/// iOS binding for Cheetah speech-to-text engine. Provides a Swift interface to the Cheetah library.
public class Cheetah {
    private static let PV_API_URL = "https://rest.picovoice.ai/"
    private static let VALID_LANGUAGES: Set<String> = ["en", "fr", "de", "es", "it", "pt"]

    private var handle: OpaquePointer?

    /// The number of audio samples per frame.
    public static let frameLength = UInt32(pv_cheetah_frame_length())

    /// Audio sample rate accepted by Cheetah.
    public static let sampleRate = UInt32(pv_sample_rate())

    /// Current Cheetah version.
    public static let version = String(cString: pv_cheetah_version())

    private static var sdk = "ios"

    public static func setSdk(sdk: String) {
        self.sdk = sdk
    }

    /// Constructor.
    ///
    /// - Parameters:
    ///   - accessKey: The AccessKey obtained from Picovoice Console (https://console.picovoice.ai).
    ///   - modelPath: Absolute path to file containing model parameters.
    ///   - device: String representation of the device (e.g., CPU or GPU) to use. If set to `best`, the most
    ///   suitable device is selected automatically. If set to `gpu`, the engine uses the first available GPU
    ///   device. To select a specific GPU device, set this argument to `gpu:${GPU_INDEX}`, where `${GPU_INDEX}`
    ///   is the index of the target GPU. If set to `cpu`, the engine will run on the CPU with the default
    ///   number of threads. To specify the number of threads, set this argument to `cpu:${NUM_THREADS}`,
    ///   where `${NUM_THREADS}` is the desired number of threads.
    ///   - endpointDuration: Duration of endpoint in seconds. A speech endpoint is detected when there is a
    ///     chunk of audio (with a duration specified herein) after an utterance without any speech in it.
    ///     Set duration to 0 to disable this. Default is 1 second.
    ///   - enableAutomaticPunctuation: Set to `true` to enable automatic punctuation insertion.
    ///   - enableTextNormalization: Set to `true` to enable text normalization. Enabling this feature improves the
    ///   readability and formatting of Cheetah's transcriptions (e.g. converts number words to digits) at the cost of
    ///   some additional latency.
    /// - Throws: CheetahError
    public init(
            accessKey: String,
            modelPath: String,
            device: String? = nil,
            endpointDuration: Float = 1.0,
            enableAutomaticPunctuation: Bool = false,
            enableTextNormalization: Bool = false) throws {

        if accessKey.count == 0 {
            throw CheetahInvalidArgumentError("AccessKey is required for Cheetah initialization")
        }

        var modelPathArg = modelPath
        if !FileManager().fileExists(atPath: modelPathArg) {
            modelPathArg = try getResourcePath(modelPathArg)
        }

        var deviceArg = device
        if device == nil {
            deviceArg = "best"
        }

        if endpointDuration < 0 {
            throw CheetahInvalidArgumentError("EndpointDuration must be a non-negative number.")
        }

        pv_set_sdk(Cheetah.sdk)

        let status = pv_cheetah_init(
                accessKey,
                modelPathArg,
                deviceArg,
                endpointDuration,
                enableAutomaticPunctuation,
                enableTextNormalization,
                &handle)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah init failed", messageStack)
        }
    }

    deinit {
        self.delete()
    }

    /// Constructor.
    ///
    /// - Parameters:
    ///   - accessKey: The AccessKey obtained from Picovoice Console (https://console.picovoice.ai).
    ///   - modelURL: URL to file containing model parameters.
    ///   - endpointDuration: Duration of endpoint in seconds. A speech endpoint is detected when there is a
    ///     chunk of audio (with a duration specified herein) after an utterance without any speech in it.
    ///     Set duration to 0 to disable this. Default is 1 second.
    ///   - enableAutomaticPunctuation: Set to `true` to enable automatic punctuation insertion.
    ///   - enableTextNormalization: Set to `true` to enable text normalization. Enabling this feature improves the
    ///   readability and formatting of Cheetah's transcriptions (e.g. converts number words to digits) at the cost of
    ///   some additional latency.
    /// - Throws: CheetahError
    public convenience init(
            accessKey: String,
            modelURL: URL,
            device: String? = nil,
            endpointDuration: Float = 1.0,
            enableAutomaticPunctuation: Bool = false,
            enableTextNormalization: Bool = false) throws {
        try self.init(
                accessKey: accessKey,
                modelPath: modelURL.path,
                device: device,
                endpointDuration: endpointDuration,
                enableAutomaticPunctuation: enableAutomaticPunctuation,
                enableTextNormalization: enableTextNormalization)
    }

    /// Releases native resources that were allocated to Cheetah
    public func delete() {
        if handle != nil {
            pv_cheetah_delete(handle)
            handle = nil
        }
    }

    /// Processes a frame of audio and returns newly-transcribed text and a flag indicating if an endpoint has been
    /// detected. Upon detection of an endpoint, the client may invoke `.flush()` to retrieve any remaining
    /// transcription.
    ///
    /// - Parameters:
    ///   - pcm: A frame of audio samples. The number of samples per frame can be attained by calling
    ///     `Cheetah.frame_length`. The incoming audio needs to have a sample rate equal to `Cheetah.sample_rate`
    ///      and be 16-bit linearly-encoded. Furthermore, Cheetah operates on single-channel audio.
    /// - Throws: CheetahError
    /// - Returns: Tuple of any newly-transcribed speech (if none is available then an empty string is returned) and a
    ///   flag indicating if an endpoint has been detected.
    public func process(_ pcm: [Int16]) throws -> (String, Bool) {
        if handle == nil {
            throw CheetahInvalidStateError("Cheetah must be initialized before processing")
        }

        if pcm.count != Cheetah.frameLength {
            throw CheetahInvalidArgumentError(
                "Frame of audio data must contain \(Cheetah.frameLength) samples" +
                " - given frame contained \(pcm.count)")
        }

        var cPartialTranscript: UnsafeMutablePointer<Int8>?
        var endPoint: Bool = false
        var cNumWords: Int32 = 0
        var cWords: UnsafeMutablePointer<pv_word_t>?
        let status = pv_cheetah_process(
                self.handle,
                pcm,
                &cPartialTranscript,
                &cNumWords,
                &cWords,
                &endPoint)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah process failed", messageStack)
        }

        let transcript = String(cString: cPartialTranscript!)
        pv_cheetah_transcript_delete(cPartialTranscript!)
        if cWords != nil {
            pv_cheetah_words_delete(cNumWords, cWords!)
        }

        return (transcript, endPoint)
    }

    /// Marks the end of the audio stream, flushes internal state of the object, and returns
    /// any remaining transcribed text.
    ///
    /// - Returns: Any remaining transcribed text. If none is available then an empty string is returned.
    public func flush() throws -> String {
        if handle == nil {
            throw CheetahInvalidStateError("Cheetah must be initialized before flushing")
        }

        var cFinalTranscript: UnsafeMutablePointer<Int8>?
        var cNumWords: Int32 = 0
        var cWords: UnsafeMutablePointer<pv_word_t>?
        let status = pv_cheetah_flush(self.handle, &cFinalTranscript, &cNumWords, &cWords)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah flush failed", messageStack)
        }

        let transcript = String(cString: cFinalTranscript!)
        pv_cheetah_transcript_delete(cFinalTranscript!)
        if cWords != nil {
            pv_cheetah_words_delete(cNumWords, cWords!)
        }

        return transcript
    }

    /// Processes a frame of audio and returns newly-transcribed text and a flag indicating if an endpoint has been
    /// detected. Upon detection of an endpoint, the client may invoke `.flushAnnotated()` to retrieve any remaining
    /// transcription.
    ///
    /// - Parameters:
    ///   - pcm: A frame of audio samples. The number of samples per frame can be attained by calling
    ///     `Cheetah.frame_length`. The incoming audio needs to have a sample rate equal to `Cheetah.sample_rate`
    ///      and be 16-bit linearly-encoded. Furthermore, Cheetah operates on single-channel audio.
    /// - Throws: CheetahError
    /// - Returns: A CheetahTranscriptAnnotated object containing any newly-transcribed speech, a flag indicating if an
    ///   endpoint has been detected, and a list of transcribed words.
    public func processAnnotated(_ pcm: [Int16]) throws -> CheetahTranscriptAnnotated {
        if handle == nil {
            throw CheetahInvalidStateError("Cheetah must be initialized before processing")
        }

        if pcm.count != Cheetah.frameLength {
            throw CheetahInvalidArgumentError(
                "Frame of audio data must contain \(Cheetah.frameLength) samples" +
                " - given frame contained \(pcm.count)")
        }

        var cPartialTranscript: UnsafeMutablePointer<Int8>?
        var endPoint: Bool = false
        var cNumWords: Int32 = 0
        var cWords: UnsafeMutablePointer<pv_word_t>?
        let status = pv_cheetah_process(
                self.handle,
                pcm,
                &cPartialTranscript,
                &cNumWords,
                &cWords,
                &endPoint)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah process failed", messageStack)
        }

        let transcript = String(cString: cPartialTranscript!)
        pv_cheetah_transcript_delete(cPartialTranscript!)

        var words: [CheetahWord] = []
        if cWords != nil {
            for i in 0..<Int(cNumWords) {
                words.append(CheetahWord(
                    word: String(cString: cWords![i].word),
                    startSec: cWords![i].start_sec,
                    endSec: cWords![i].end_sec,
                    confidence: cWords![i].confidence))
            }
            pv_cheetah_words_delete(cNumWords, cWords!)
        }

        return CheetahTranscriptAnnotated(
            transcript: transcript,
            isEndpoint: endPoint,
            words: words)
    }

    /// Marks the end of the audio stream, flushes internal state of the object, and returns
    /// any remaining transcribed text.
    ///
    /// - Returns: Returns: A CheetahTranscriptAnnotated object containing any newly-transcribed speech, and a list of
    ///   transcribed words.
    public func flushAnnotated() throws -> CheetahTranscriptAnnotated {
        if handle == nil {
            throw CheetahInvalidStateError("Cheetah must be initialized before flushing")
        }

        var cFinalTranscript: UnsafeMutablePointer<Int8>?
        var cNumWords: Int32 = 0
        var cWords: UnsafeMutablePointer<pv_word_t>?
        let status = pv_cheetah_flush(self.handle, &cFinalTranscript, &cNumWords, &cWords)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah flush failed", messageStack)
        }

        let transcript = String(cString: cFinalTranscript!)
        pv_cheetah_transcript_delete(cFinalTranscript!)

        var words: [CheetahWord] = []
        if cWords != nil {
            for i in 0..<Int(cNumWords) {
                words.append(CheetahWord(
                    word: String(cString: cWords![i].word),
                    startSec: cWords![i].start_sec,
                    endSec: cWords![i].end_sec,
                    confidence: cWords![i].confidence))
            }
            pv_cheetah_words_delete(cNumWords, cWords!)
        }

        return CheetahTranscriptAnnotated(
            transcript: transcript,
            isEndpoint: false,
            words: words)
    }

    /// Given a path, return the full path to the resource.
    ///
    /// - Parameters:
    ///   - filePath: relative path of a file in the bundle.
    /// - Throws: LeopardIOError
    /// - Returns: The full path of the resource.
    private func getResourcePath(_ filePath: String) throws -> String {
        if let resourcePath = Bundle(for: type(of: self)).resourceURL?.appendingPathComponent(filePath).path {
            if FileManager.default.fileExists(atPath: resourcePath) {
                return resourcePath
            }
        }

        throw CheetahIOError(
            "Could not find file at path '\(filePath)'. If this is a packaged asset, " +
            "ensure you have added it to your xcode project."
        )
    }

    /// Trains a new model using the specified `newWords` and `boostWords` arguments.
    ///
    /// - Parameters:
    ///   - accessKey: AccessKey obtained from Picovoice Console (https://console.picovoice.ai/).
    ///   - outputPath: Absolute path to file where the trained model will be saved.
    ///   - language: Two character language code for the model (e.g. "en", "fr").
    ///               See https://picovoice.ai/docs/model-api/cheetah/ for supported languages.
    ///   - newWords: A dictionary of words to pronunciations to add to the new model. Keys should be
    ///               the word string. Values are a Set of pronunciations for the given word, each pronunciation
    ///               is a string of space separated IPA phonemes. An empty Set will result in the training
    ///               generating a default pronunciation.
    ///   - boostWords: A Set of words to "boost". The engine will be more likely to select the boosted words.

    /// - Throws: `CheetahError` if model training fails.
    public static func trainModelFromWords(
        accessKey: String,
        outputPath: String,
        language: String,
        newWords: [String: Set<String>],
        boostWords: Set<String>
    ) throws {
        var newWordsContent: [String: [String]] = [:]
        for (key, value) in newWords {
            newWordsContent[key] = Array(value)
        }
        let boostWordsContent: [String] = Array(boostWords)
        
        let root: [String: Any] = [
            "new": newWords,
            "boost": boostWords
        ]

        let newYaml: String
        do {
            newYaml = try Yams.dump(object: root)
        } catch {
            throw CheetahRuntimeError("Failed to serialize yaml content")
        }

        try trainModelFromYaml(
            accessKey: accessKey,
            outputPath: outputPath,
            language: language,
            yamlContent: newYaml)
    }

    /// Trains a model using a YAML configuration string.
    ///
    /// - Parameters:
    ///   - accessKey: AccessKey obtained from Picovoice Console (https://console.picovoice.ai/).
    ///   - outputPath: Absolute path to file where the trained model will be saved.
    ///   - language: Two character language code for the model (e.g. "en", "fr").
    ///               See https://picovoice.ai/docs/model-api/cheetah/ for supported languages.
    ///   - yamlContent: YAML configuration string to be used for training.
    /// - Throws: `CheetahError` if model training fails.
    public static func trainModelFromYaml(
        accessKey: String,
        outputPath: String,
        language: String,
        yamlContent: String
    ) throws {

        guard VALID_LANGUAGES.contains(language) else {
            throw CheetahInvalidArgumentError("Invalid language ('\(language)')")
        }

        let payload: Data
        do {
            payload = try JSONSerialization.data(withJSONObject: [
                "engine": "cheetah",
                "model_type": "default",
                "yaml_content": yamlContent
            ])
        } catch {
            throw CheetahRuntimeError("Failed to create request payload \(error.localizedDescription)")
        }

        guard let url = URL(string: PV_API_URL + language + "/api/cat") else {
            throw CheetahRuntimeError("Invalid request URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = payload
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(accessKey, forHTTPHeaderField: "x-api-key")

        var resultData: Data?
        var resultResponse: URLResponse?
        var resultError: Error?
        let semaphore = DispatchSemaphore(value: 0)

        URLSession.shared.dataTask(with: request) { data, response, error in
            resultData = data
            resultResponse = response
            resultError = error
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let error = resultError {
            throw CheetahRuntimeError("Request failed: \(error.localizedDescription)")
        }

        guard let http = resultResponse as? HTTPURLResponse else {
            throw CheetahRuntimeError("Invalid response")
        }

        guard (200...299).contains(http.statusCode) else {
            let errorBody = resultData.flatMap { String(data: $0, encoding: .utf8) } ?? ""
            throw CheetahRuntimeError("Failed to train model: \(errorBody)")
        }

        guard let data = resultData, !data.isEmpty else {
            throw CheetahRuntimeError("Empty response body")
        }

        do {
            try data.write(to: URL(fileURLWithPath: outputPath))
        } catch {
            throw CheetahRuntimeError("Failed to save Cheetah model file: \(error.localizedDescription)")
        }
    }

    /// Lists all available devices that Cheetah can use for inference.
    /// Entries in the list can be used as the `device` argument when initializing Cheetah.
    ///
    /// - Throws: CheetahError
    /// - Returns: Array of available devices that Cheetah can be used for inference.
    public static func getAvailableDevices() throws -> [String] {
        var cHardwareDevices: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?
        var numHardwareDevices: Int32 = 0
        let status = pv_cheetah_list_hardware_devices(&cHardwareDevices, &numHardwareDevices)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try Cheetah.getMessageStack()
            throw Cheetah.pvStatusToCheetahError(status, "Cheetah getAvailableDevices failed", messageStack)
        }

        var hardwareDevices: [String] = []
        for i in 0..<numHardwareDevices {
            hardwareDevices.append(String(cString: cHardwareDevices!.advanced(by: Int(i)).pointee!))
        }

        pv_cheetah_free_hardware_devices(cHardwareDevices, numHardwareDevices)

        return hardwareDevices
    }

    private static func pvStatusToCheetahError(
        _ status: pv_status_t,
        _ message: String,
        _ messageStack: [String] = []) -> CheetahError {
        switch status {
        case PV_STATUS_OUT_OF_MEMORY:
            return CheetahMemoryError(message, messageStack)
        case PV_STATUS_IO_ERROR:
            return CheetahIOError(message, messageStack)
        case PV_STATUS_INVALID_ARGUMENT:
            return CheetahInvalidArgumentError(message, messageStack)
        case PV_STATUS_STOP_ITERATION:
            return CheetahStopIterationError(message, messageStack)
        case PV_STATUS_KEY_ERROR:
            return CheetahKeyError(message, messageStack)
        case PV_STATUS_INVALID_STATE:
            return CheetahInvalidStateError(message, messageStack)
        case PV_STATUS_RUNTIME_ERROR:
            return CheetahRuntimeError(message, messageStack)
        case PV_STATUS_ACTIVATION_ERROR:
            return CheetahActivationError(message, messageStack)
        case PV_STATUS_ACTIVATION_LIMIT_REACHED:
            return CheetahActivationLimitError(message, messageStack)
        case PV_STATUS_ACTIVATION_THROTTLED:
            return CheetahActivationThrottledError(message, messageStack)
        case PV_STATUS_ACTIVATION_REFUSED:
            return CheetahActivationRefusedError(message, messageStack)
        default:
            let pvStatusString = String(cString: pv_status_to_string(status))
            return CheetahError("\(pvStatusString): \(message)", messageStack)
        }
    }

    private static func getMessageStack() throws -> [String] {
        var messageStackRef: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?
        var messageStackDepth: Int32 = 0
        let status = pv_get_error_stack(&messageStackRef, &messageStackDepth)
        if status != PV_STATUS_SUCCESS {
            throw Cheetah.pvStatusToCheetahError(status, "Unable to get Cheetah error state")
        }

        var messageStack: [String] = []
        for i in 0..<messageStackDepth {
            messageStack.append(String(cString: messageStackRef!.advanced(by: Int(i)).pointee!))
        }

        pv_free_error_stack(messageStackRef)

        return messageStack
    }
}

File: binding/java/src/ai/picovoice/cheetah/Cheetah.java
/*
    Copyright 2022-2025 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.cheetah;

import java.io.File;
import java.util.ArrayList;

/**
 * Cheetah Class.
 */
public class Cheetah {

    private static String sdk = "java";

    public static final String LIBRARY_PATH;
    public static final String MODEL_PATH;

    static {
        LIBRARY_PATH = Utils.getPackagedLibraryPath();
        MODEL_PATH = Utils.getPackagedModelPath();
    }

    public static void setSdk(String sdk) {
        Cheetah.sdk = sdk;
    }

    private long handle;

    /**
     * Constructor.
     *
     * @param accessKey                  AccessKey obtained from Picovoice Console.
     * @param modelPath                  Absolute path to the file containing model parameters.
     * @param device                     String representation of the device (e.g., CPU or GPU) to use. If set to `best`, 
     *                                   the most suitable device is selected automatically. If set to `gpu`, the engine
     *                                   uses the first available GPU device. To select a specific GPU device, set this
     *                                   argument to `gpu:${GPU_INDEX}`, where `${GPU_INDEX}` is the index of the target
     *                                   GPU. If set to `cpu`, the engine will run on the CPU with the default number of
     *                                   threads. To specify the number of threads, set this argument to
     *                                   `cpu:${NUM_THREADS}`, where `${NUM_THREADS}` is the desired number of threads.
     * @param libraryPath                Absolute path to the native Cheetah library.
     * @param endpointDurationSec        Duration of endpoint in seconds. A speech endpoint is detected when there is a
     *                                   chunk of audio (with a duration specified herein) after an utterance without
     *                                   any speech in it. Set duration to 0 to disable this. 
     *                                   Default is 1 second in the Builder.
     * @param enableAutomaticPunctuation Set to `true` to enable automatic punctuation insertion.
     * @throws CheetahException if there is an error while initializing Cheetah.
     */
    private Cheetah(
            String accessKey,
            String modelPath,
            String device,
            String libraryPath,
            float endpointDurationSec,
            boolean enableAutomaticPunctuation) throws CheetahException {
        try {
            ArrayList<String> libraryDependencies = Utils.getLibraryDependencyPaths(libraryPath);
            for (String dependency : libraryDependencies) {
                System.load(dependency);
            }
            System.load(libraryPath);
        } catch (Exception exception) {
            throw new CheetahException(exception);
        }
        CheetahNative.setSdk(Cheetah.sdk);

        handle = CheetahNative.init(
                accessKey,
                modelPath,
                device,
                endpointDurationSec,
                enableAutomaticPunctuation);
    }

    /**
     * Releases resources acquired by Cheetah.
     */
    public void delete() {
        if (handle != 0) {
            CheetahNative.delete(handle);
            handle = 0;
        }
    }

    /**
     * Processes given audio data and returns its transcription.
     *
     * @param pcm A frame of audio samples. The number of samples per frame can be attained by
     *            calling {@link #getFrameLength()}. The incoming audio needs to have a sample rate
     *            equal to {@link #getSampleRate()} and be 16-bit linearly-encoded. Furthermore,
     *            Cheetah operates on single channel audio only.
     * @return Inferred transcription.
     * @throws CheetahException if there is an error while processing the audio frame.
     */
    public CheetahTranscript process(short[] pcm) throws CheetahException {
        if (handle == 0) {
            throw new CheetahInvalidStateException("Attempted to call Cheetah process after delete.");
        }

        if (pcm == null) {
            throw new CheetahInvalidArgumentException("Passed null frame to Cheetah process.");
        }

        if (pcm.length != getFrameLength()) {
            throw new CheetahInvalidArgumentException(
                    String.format("Cheetah process requires frames of length %d. " +
                            "Received frame of size %d.", getFrameLength(), pcm.length));
        }
        return CheetahNative.process(handle, pcm);
    }

    /**
     * Processes any remaining audio data and returns its transcription.
     *
     * @return Inferred transcription.
     * @throws CheetahException if there is an error while processing the audio frame.
     */
    public CheetahTranscript flush() throws CheetahException {
        if (handle == 0) {
            throw new CheetahInvalidStateException("Attempted to call Cheetah flush after delete.");
        }
        return CheetahNative.flush(handle);
    }

    /**
     * Getter for required number of audio samples per frame.
     *
     * @return Required number of audio samples per frame.
     */
    public int getFrameLength() {
        return CheetahNative.getFrameLength();
    }

    /**
     * Getter for required audio sample rate for PCM data.
     *
     * @return Required audio sample rate for PCM data.
     */
    public int getSampleRate() {
        return CheetahNative.getSampleRate();
    }

    /**
     * Getter for Cheetah version.
     *
     * @return Cheetah version.
     */
    public String getVersion() {
        return CheetahNative.getVersion();
    }

    /**
     * Retrieves a list of available hardware devices that Cheetah can use to run inference.
     *
     * @param libraryPath Path to a native Cheetah library. Set to `null` to use default library.
     *
     * @return List of available hardware devices that Cheetah can use to run inference.
     * @throws CheetahException if the library file cannot be loaded.
     */
    public static String[] getAvailableDevices(String libraryPath) throws CheetahException {
        try {
            System.load(libraryPath);
        } catch (Exception exception) {
            throw new CheetahException(exception);
        }
        return CheetahNative.listHardwareDevices();
    }

    /**
     * Retrieves a list of available hardware devices that Cheetah can use to run inference.
     *
     * @return List of available hardware devices that Cheetah can use to run inference.
     * @throws CheetahException if the default library file cannot be loaded.
     */
    public static String[] getAvailableDevices() throws CheetahException {
        if (Utils.isResourcesAvailable()) {
            return Cheetah.getAvailableDevices(LIBRARY_PATH);
        } else {
            throw new CheetahInvalidArgumentException("Default library unavailable. " +
                    "Please provide a valid native Cheetah library path.");
        }
    }

    /**
     * Builder for creating an instance of Cheetah with a mixture of default arguments.
     */
    public static class Builder {
        private String accessKey = null;
        private String libraryPath = null;
        private String device = null;
        private String modelPath = null;
        private float endpointDuration = 1f;
        private boolean enableAutomaticPunctuation = false;

        public Builder setAccessKey(String accessKey) {
            this.accessKey = accessKey;
            return this;
        }

        public Builder setLibraryPath(String libraryPath) {
            this.libraryPath = libraryPath;
            return this;
        }

        /**
         * Setter for the absolute path to the file containing Cheetah model parameters.
         *
         * @param modelPath Absolute path to the file containing Cheetah model parameters.
         */
        public Builder setModelPath(String modelPath) {
            this.modelPath = modelPath;
            return this;
        }

        public Builder setDevice(String device) {
            this.device = device;
            return this;
        }

        /**
         * Setter for the duration of endpoint in seconds.
         *
         * @param endpointDuration Duration of endpoint in seconds. A speech endpoint is detected when there is a
         *                         chunk of audio (with a duration specified herein) after an utterance without
         *                         any speech in it.
         */
        public Builder setEndpointDuration(float endpointDuration) {
            this.endpointDuration = endpointDuration;
            return this;
        }

        /**
         * Setter for enabling automatic punctuation insertion.
         *
         * @param enableAutomaticPunctuation Set to `true` to enable automatic punctuation insertion.
         */
        public Builder setEnableAutomaticPunctuation(boolean enableAutomaticPunctuation) {
            this.enableAutomaticPunctuation = enableAutomaticPunctuation;
            return this;
        }

        /**
         * Validates properties and creates an instance of the Cheetah speech-to-text engine.
         *
         * @return An instance of Cheetah Engine
         * @throws CheetahException if there is an error while initializing Cheetah.
         */
        public Cheetah build() throws CheetahException {
            if (!Utils.isEnvironmentSupported()) {
                throw new CheetahRuntimeException("Could not initialize Cheetah. " +
                        "Execution environment not currently supported by Cheetah Java.");
            }

            if (accessKey == null) {
                throw new CheetahInvalidArgumentException("AccessKey must not be null");
            }

            if (device == null) {
                device = "best";
            }

            if (libraryPath == null) {
                if (Utils.isResourcesAvailable()) {
                    libraryPath = LIBRARY_PATH;
                } else {
                    throw new CheetahInvalidArgumentException("Default library unavailable. Please " +
                            "provide a native Cheetah library path (-l <library_path>).");
                }
                if (!new File(libraryPath).exists()) {
                    throw new CheetahIOException(String.format("Couldn't find library file at " +
                            "'%s'", libraryPath));
                }
            }

            if (modelPath == null) {
                if (Utils.isResourcesAvailable()) {
                    modelPath = MODEL_PATH;
                } else {
                    throw new CheetahInvalidArgumentException("Default model unavailable. Please provide a " +
                            "valid Cheetah model path (-m <model_path>).");
                }
                if (!new File(modelPath).exists()) {
                    throw new CheetahIOException(String.format("Couldn't find model file at " +
                            "'%s'", modelPath));
                }
            }

            if (endpointDuration < 0f) {
                throw new CheetahInvalidArgumentException("endpointDuration must be greater than or equal to 0.0");
            }

            return new Cheetah(
                    accessKey,
                    modelPath,
                    device,
                    libraryPath,
                    endpointDuration,
                    enableAutomaticPunctuation);
        }
    }
}
